flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/WaitingForResources.java

```diff
@@ -90,13 +107,40 @@ public Logger getLogger() {
 
     @Override
     public void notifyNewResourcesAvailable() {
-        if (context.hasEnoughResources(desiredResources)) {
+        checkDesiredOrSufficientResourcesAvailable();
+    }
+
+    private void checkDesiredOrSufficientResourcesAvailable() {
+        if (context.hasDesiredResources(desiredResources)) {
             createExecutionGraphWithAvailableResources();
+            return;
         }
+
+        if (context.hasSufficientResources()) {
+            Deadline deadline = initializeOrGetResourceStabilizationDeadline();
+            if (deadline.isOverdue()) {
+                createExecutionGraphWithAvailableResources();
+            } else {
+                // schedule next resource check
+                context.runIfState(
+                        this,
+                        this::checkDesiredOrSufficientResourcesAvailable,
+                        deadline.timeLeft());
+            }
+        }
```

Comment:
Shouldn't we clear the resource stabilization deadline if we no longer have sufficient resources so that we can start a new deadline once we have sufficient resources again?